Support writing cluster config parser output to a file

cshannon opened the pull request:
- This commit adds support to provide an optional argument to ClusterConfigParser to write the output to a file instead of System.out so that only the output we want is captured and not everything that's written to standard out.
-  This the accumulo-cluster script has also been updated to to provide the temp file as an output file argument instead of capturing System.out so that there are not parsing errors by the script if other output is written to the standard out stream.

This fixes parsing issues if there's anything that writes to standard out/error but the main reason for this PR is to fix https://github.com/apache/accumulo-classloaders/issues/18 because the debug output (or any other output) by the classloader to standard out/err will no longer interfere with the parsing by the accumulo-cluster script.

cshannon commented:
I added a unit test and I also tested using [ReloadingVFSContextClassLoaderFactory](https://github.com/apache/accumulo-classloaders/blob/765db685f362ef38a276573a012769ca474f5d9c/modules/vfs-class-loader/src/main/java/org/apache/accumulo/classloader/vfs/context/ReloadingVFSContextClassLoaderFactory.java) in accumulo-classloaders against Uno with the the factory configured as the system classloader. Now with debug logging enabled everything works as intended.

dlmarion commented on core/src/main/java/org/apache/accumulo/core/conf/cluster/ClusterConfigParser.java:
```diff
@@ -135,12 +136,23 @@ public static void outputShellVariables(Map<String,String> config, PrintStream o
-    if (args == null || args.length != 1) {
-      System.err.println("Usage: ClusterConfigParser <configFile>");
+    if (args == null || args.length < 1 || args.length > 2) {
+      System.err.println("Usage: ClusterConfigParser <configFile> [<outputFile]");
```
```suggestion
      System.err.println("Usage: ClusterConfigParser <configFile> [<outputFile>]");
```

dlmarion reviewed (approved):
LGTM, small nit

dlmarion commented:
I'm thinking that we may want to target this for 2.1 vs main.

cshannon commented:
> I'm thinking that we may want to target this for 2.1 vs main.

Ah yeah, good point, this should likely go into 2.1 as well because I view this is a bug since anything that writes to standard out (not just the VFS classloader) while parsing can break it. I'll switch it.

cshannon commented:
@dlmarion - I fixed that one typo you found and I went ahead and just squashed my commits and force pushed off 2.1. I could have just switched the target branch but I wanted to make sure everything was good when starting from 2.1 so I just created a new branch from 2.1 and applied my changes to it with a squashed commit and pushed it back up.

dlmarion reviewed (approved):
Feel free to update the 2.1.1 release notes in the website repo post-merge.

ctubbsii commented on assemble/bin/accumulo-cluster:
```diff
@@ -66,7 +66,7 @@ function parse_config {
   trap 'rm -f "$CONFIG_FILE"' EXIT
   CONFIG_FILE=$(mktemp) || exit 1
-  ${accumulo_cmd} org.apache.accumulo.core.conf.cluster.ClusterConfigParser "${conf}"/cluster.yaml >"$CONFIG_FILE" || parse_fail
+  ${accumulo_cmd} org.apache.accumulo.core.conf.cluster.ClusterConfigParser "${conf}"/cluster.yaml "$CONFIG_FILE" || parse_fail
```
Instead of just relying on the number of args, it would be better to have an explicit flag like `-o outputfile`. However, this is an internal tool, not public API, so it really doesn't matter as long as it works for us. I mention this here, because I think it's a good principle to keep in mind for any similar changes in future that might apply to a user facing command.
